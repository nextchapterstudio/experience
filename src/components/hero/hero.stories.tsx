import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { Hero, HeroBody, HeroFooter, HeroHead } from './'
import { Title, SubTitle } from '../'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/Hero',
  component: Hero,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof Hero>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Hero> = (args) => <Hero {...args} />

export const BasicTemplate = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
BasicTemplate.args = {
  children: 'Simple Hero',
}

const FullHeroTemplate: ComponentStory<typeof Hero> = (args) => (
  <Hero {...args}>
    <HeroHead>The Head</HeroHead>
    <HeroBody>
      <div>
        <Title as="p">The Hero</Title>
        <SubTitle as="p">We Need</SubTitle>
      </div>
    </HeroBody>
    <HeroFooter>Foot</HeroFooter>
  </Hero>
)

export const FullHeroExample = FullHeroTemplate.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
FullHeroExample.args = {
  isPrimary: true,
  isFullHeight: true,
}
