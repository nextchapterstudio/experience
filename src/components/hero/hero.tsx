import React, { createContext, FC, ReactChild, ReactElement, ReactNode } from 'react'
import classnames from 'classnames'
import { Colors, Sizes, Screens } from '../../utils/modifiers'

export interface HeroBaseProps {
  children?: ReactNode
  className?: string
}
export interface IHeroContext {
  isInHero:boolean
}
export const HeroHead: FC<HeroBaseProps> = ({ children, className }) => {
  return <div className={classnames('hero-head', className)}>{children}</div>
}
export const HeroBody: FC<HeroBaseProps>  = ({ children, className }) => {
  return <div className={classnames('hero-body', className)}>{children}</div>
}
export const HeroFooter: FC<HeroBaseProps>  = ({ children, className }) => {
  return <div className={classnames('hero-foot', className)}>{children}</div>
}
export const HeroContext = createContext<IHeroContext>({isInHero:false})

export interface HeroProps {
  isPrimary?:boolean
  isSuccess?:boolean
  isInfo?:boolean
  isWarning?:boolean
  isDanger?:boolean
  isLink?:boolean
  isLight?:boolean
  isDark?:boolean
  isMedium?:boolean
  isLarge?:boolean
  isBold?:boolean
  isFullHeight?:boolean
  isFullHeightWithNavBar?:boolean
}
export const Hero: FC<HeroBaseProps & HeroProps>  = ({
  children,
  className,
  isPrimary,
  isSuccess,
  isInfo,
  isWarning,
  isDanger,
  isLink,
  isLight,
  isDark,
  isMedium,
  isLarge,
  isBold,
  isFullHeight,
  isFullHeightWithNavBar,
}) => {
  const classes = Colors({
    isPrimary,
    isSuccess,
    isInfo,
    isWarning,
    isDanger,
    isLink,
    isLight,
    isDark,
  })
  const sizeClasses = Sizes({
    isMedium,
    isLarge,
  })
  const screenSizes = Screens({
    isFullHeight,
    isFullHeightWithNavBar,
  })
  const kid = children as ReactElement
  const kids = React.Children.toArray(children)
  console.log('kid?.type', kid?.type)
  console.log('head', kid?.type !== HeroHead)

  if (kids.length === 1 && (kid?.type !== HeroBody && kid?.type !== HeroHead && kid?.type !== HeroFooter)) {
    console.log('singly????', kid?.type)
    return (
      <HeroContext.Provider value={{ isInHero: true }}>
        <section
          className={classnames(
            'hero',
            className,
            classes,
            sizeClasses,
            screenSizes,
            {
              'is-bold': isBold,
            }
          )}
        >
          <div className="hero-body">{children}</div>
        </section>
      </HeroContext.Provider>
    )
  }


  return (
    <HeroContext.Provider value={{ isInHero: true }}>
      <section
        className={classnames(
          'hero',
          className,
          classes,
          sizeClasses,
          screenSizes,
          {
            'is-bold': isBold,
          }
        )}
      >
        {kids}
      </section>
    </HeroContext.Provider>
  )
}
