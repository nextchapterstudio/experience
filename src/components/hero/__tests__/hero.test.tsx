import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import { Hero, HeroBody, HeroFooter, HeroHead } from '../hero'

describe('The Hero Component', () => {
  afterEach(cleanup)
  it('should render A Full Hero with just the hero', () => {
    const { container } = render(<Hero>test</Hero>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('hero')
    expect(container.firstChild?.firstChild).toHaveClass('hero-body')
  })
  it('should render A Hero with the head element', () => {
    const { container } = render(<Hero><HeroHead>The Head</HeroHead></Hero>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('hero')
    expect(container.firstChild?.firstChild).toHaveClass('hero-head')
  })
  it('should render A Hero with the head body and footer', () => {
    const { container } = render(<Hero><HeroHead>The Head</HeroHead><HeroBody>Body</HeroBody><HeroFooter>Foot</HeroFooter></Hero>)

    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('hero')
    expect(container.firstChild?.firstChild).toHaveClass('hero-head')
    expect(container.firstChild?.firstChild?.nextSibling).toHaveClass('hero-body')
    expect(container.firstChild?.firstChild?.nextSibling?.nextSibling).toHaveClass('hero-foot')
  })
})
