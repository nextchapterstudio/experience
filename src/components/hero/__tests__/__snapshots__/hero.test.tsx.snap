// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Hero Component should render A Full Hero with just the hero 1`] = `
<div>
  <section
    class="hero"
  >
    <div
      class="hero-body"
    >
      test
    </div>
  </section>
</div>
`;

exports[`The Hero Component should render A Full Hero with just the hero 2`] = `
<section
  class="hero"
>
  <div
    class="hero-body"
  >
    test
  </div>
</section>
`;

exports[`The Hero Component should render A Hero with the head body and footer 1`] = `
<section
  class="hero"
>
  <div
    class="hero-head"
  >
    The Head
  </div>
  <div
    class="hero-body"
  >
    Body
  </div>
  <div
    class="hero-foot"
  >
    Foot
  </div>
</section>
`;

exports[`The Hero Component should render A Hero with the head element 1`] = `
<div>
  <section
    class="hero"
  >
    <div
      class="hero-head"
    >
      The Head
    </div>
  </section>
</div>
`;

exports[`The Hero Component should render A Hero with the head element 2`] = `
<section
  class="hero"
>
  <div
    class="hero-head"
  >
    The Head
  </div>
</section>
`;
